src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py

```diff
@@ -742,9 +746,14 @@ def remove_snapshot(self, snapname):
         if self.has_pending_clones(snapname):
             raise VolumeException(-errno.EAGAIN, "snapshot '{0}' has pending clones".format(snapname))
         snappath = self.snapshot_path(snapname)
+        try:
+            self.metadata_mgr.remove_section(self.get_snap_section_name(snapname))
```

Comment:
@vshankar @nmshelke @ajarr 

The order is reversed here. Previously, snapshot was removed first and then the metadata, now the other way. This is the right thing to do as we  end up keeping the metadata of the deleted snapshot if metadata deletion fails. 

But there is a side effect of doing this. The metadata deletion usually fails with ENOSPACE (OSD full scenario). So when there is no space, the snapshot deletion fails with new code. But with old code, the snapshot deletion is successful but the metadata deletion fails. Though the older behavior is not correct, it helps in recovering space during OSD full scenario through subvolume APIs.

We need to think and come up with some sort of solution to make snapshot deletion goes through when there is no space.  One solution is to allow the older behavior with `force` option. But we have to think of a solution to clean up the stale snapshot metadata. 

Thought ?

Reply:
I recall that we discussed this, did we decide on keeping the old behavior (delete snapshot first) and if the metadata update fails, we still report success and later when loading the subvolume, cleanup the metadata for which snaps are not present but its metadata exist.

Reply:
Yes, we discussed that in one to one meeting. We had discussed to keep the old behavior with `force` option. It's not exactly deleting the snapshot first but ignoring the metadata update failure and reporting success.

>  later when loading the subvolume, cleanup the metadata for which snaps are not present but its metadata exist?
Is this scalable ? If there are thousands of snapshots per subvolume, we end up stating thousand+ times for every such subvolume load.

Reply:
> Yes, we discussed that in one to one meeting. We had discussed to keep the old behavior with `force` option. It's not exactly deleting the snapshot first but ignoring the metadata update failure and reporting success.
> 
> > later when loading the subvolume, cleanup the metadata for which snaps are not present but its metadata exist?
> 
> Is this scalable ? If there are thousands of snapshots per subvolume, we end up stating thousand+ times for every such subvolume load.

I recall now -- we thought about maintaining an index but that wasn't feasible either (w/ ENOSPC).

Reply:
> @vshankar @nmshelke @ajarr
> 
> The order is reversed here. Previously, snapshot was removed first and then the metadata, now the other way. This is the right thing to do as we end up keeping the metadata of the deleted snapshot if metadata deletion fails.
> 
> But there is a side effect of doing this. The metadata deletion usually fails with ENOSPACE (OSD full scenario). So when there is no space, the snapshot deletion fails with new code. But with old code, the snapshot deletion is successful but the metadata deletion fails. Though the older behavior is not correct, it helps in recovering space during OSD full scenario through subvolume APIs.

This is more and more pushing me to think about ditching .meta and putting the metadata in a actual db (sqlite).

> 
> We need to think and come up with some sort of solution to make snapshot deletion goes through when there is no space. One solution is to allow the older behavior with `force` option. But we have to think of a solution to clean up the stale snapshot metadata.
> 
> Thought ?

Reply:
> > Yes, we discussed that in one to one meeting. We had discussed to keep the old behavior with `force` option. It's not exactly deleting the snapshot first but ignoring the metadata update failure and reporting success.
> > > later when loading the subvolume, cleanup the metadata for which snaps are not present but its metadata exist?
> > 
> > 
> > Is this scalable ? If there are thousands of snapshots per subvolume, we end up stating thousand+ times for every such subvolume load.
> 
> I don't think we need to do a stat; can't we just list the .snap directory and then compare to the snap names? That should scale just fine.


@vshankar @gregsfortytwo 
Listing the .snap directory is time consuming on scale. I increased the `mds_max_snaps_per_dir` to maximum allowed i.e., 4096 . The listdir of 4096 snapshots on vstart cluster took 20+ seconds. If we choose to clean up stale snap metadata in subvolume discover, every subvolume API would consume so much time which is unacceptable.

Well, it works well with the default `mds_max_snaps_per_dir` i.e., 100. Listing 100 snaps is fast enough.

I am not really sure on what's the avg number of snapshots per subvolume in general but the solution to fix in subvolume discover doesn't scale well with allowed max snapshot limit.

How about leaving the stale snapshot metadata to stay around and clean it up on snapshot creation with the same name ? This way, the stale snapshot metadata stays and goes away in following two cases.

1.  When the subvolume is deleted
2.  When the snapshot is created with the same name in the corresponding subvolume.